src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.Http2KeepAlivePing.cs

```diff
@@ -0,0 +1,371 @@
+    {
+        public static readonly bool IsSupported = PlatformDetection.SupportsAlpn && PlatformDetection.IsNotBrowser;
+
+        protected override Version UseVersion => HttpVersion20.Value;
+
+        private int _pingCounter;
+        private Http2LoopbackConnection _connection;
+        private SemaphoreSlim _writeSemaphore = new SemaphoreSlim(1);
+        private Channel<Frame> _framesChannel = Channel.CreateUnbounded<Frame>();
+        private CancellationTokenSource _incomingFramesCts = new CancellationTokenSource();
+        private Task _incomingFramesTask;
+        private TaskCompletionSource _serverFinished = new TaskCompletionSource();
+        private int _sendPingResponse = 1;
+
+        private static Http2Options NoAutoPingResponseHttp2Options => new Http2Options() { EnableTransparentPingResponse = false };
+
+        private static readonly TimeSpan TestTimeout = TimeSpan.FromSeconds(60);
+
+        public SocketsHttpHandler_Http2KeepAlivePing_Test(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [OuterLoop("Runs long")]
+        [Fact]
+        public async Task KeepAlivePingDelay_Infinite_NoKeepAlivePingIsSent()
+        {
+            await Http2LoopbackServer.CreateClientAndServerAsync(async uri =>
+            {
+                SocketsHttpHandler handler = new SocketsHttpHandler()
+                {
+                    KeepAlivePingTimeout = TimeSpan.FromSeconds(1),
+                    KeepAlivePingPolicy = HttpKeepAlivePingPolicy.Always,
+                    KeepAlivePingDelay = Timeout.InfiniteTimeSpan
+                };
+                handler.SslOptions.RemoteCertificateValidationCallback = delegate { return true; };
+
+                using HttpClient client = new HttpClient(handler);
+                client.DefaultRequestVersion = HttpVersion.Version20;
+
+                // Warmup request to create connection:
+                await client.GetStringAsync(uri);
+
+                // Actual request:
+                await client.GetStringAsync(uri);
+
+                // Let connection live until server finishes:
+                await _serverFinished.Task.WaitAsync(TestTimeout);
+            },
+            async server =>
+            {
+                await EstablishConnectionAsync(server);
+
+                // Warmup the connection.
+                int streamId1 = await ReadRequestHeaderAsync();
+                await GuardConnctionWriteAsync(() => _connection.SendDefaultResponseAsync(streamId1));
+
+                Interlocked.Exchange(ref _pingCounter, 0); // reset the PING counter
+                // Request under the test scope.
+                int streamId2 = await ReadRequestHeaderAsync();
+
+                // Simulate inactive period:
+                await Task.Delay(5_000);
+
+                // We may have received one RTT PING in response to HEADERS, but should receive no KeepAlive PING
+                Assert.True(_pingCounter <= 1);
```

Comment:
Why don't we count only PINGs relevant to this test and ignore RTT ones? AFAIR, they should have different payload and be distinguishable, no?

Reply:
I wanted to avoid to take a test-dependency on such an arbitrary implementation detail.

Reply:
I actually like the idea of leveraging the different ranges for KeepAlive (>0) and RTT (<0) in tests :smile: 
We're already counting on details like heartbeat interval and timings, so this doesn't seem to me like something dangerous.